Makefile

```diff
@@ -17,154 +17,161 @@
 # * limitations under the License.
 # *
 
-TARGET_EXEC ?= hal_test
-
+# Terminal colors
 RED='\033[0;31m'
 GREEN='\033[0;32m'
 YELLOW='\033[0;33m'
 NC='\033[0m'
 ECHOE = /bin/echo -e
+MKDIR_P ?= @mkdir -p
 
-$(info $(shell ${ECHOE} ${GREEN}TARGET_EXEC [$(TARGET_EXEC)]${NC}))
+# Directories and PATH setup
+UT_CORE_DIR := $(shell dirname $(realpath $(firstword $(MAKEFILE_LIST))))
+export PATH := $(UT_CORE_DIR)/toolchain:$(PATH)
 
-UT_CORE_DIR :=$(shell dirname $(realpath $(firstword $(MAKEFILE_LIST))))
-export PATH := $(shell pwd)/toolchain:$(PATH)
+# Set target based on TARGET variable
+TARGET := $(if $(TARGET),$(TARGET),linux)
 
-# defaults for target arm
 ifeq ($(TARGET),arm)
-CUNIT_VARIANT=arm-rdk-linux-gnueabi
+CUNIT_VARIANT = arm-rdk-linux-gnueabi
 #CC := arm-rdk-linux-gnueabi-gcc -mthumb -mfpu=vfp -mcpu=cortex-a9 -mfloat-abi=soft -mabi=aapcs-linux -mno-thumb-interwork -ffixed-r8 -fomit-frame-pointer
 # CFLAGS will be overriden by Caller as required
 INC_DIRS += $(UT_CORE_DIR)/sysroot/usr/include
-TARGET = arm
 else
-TARGET = linux
+CUNIT_VARIANT = i686-pc-linux-gnu
 endif
 export TARGET
 
-$(info TARGET [$(TARGET)])
-
 # Moveable Directories based on the caller makefile
 TOP_DIR ?= $(UT_CORE_DIR)
 BIN_DIR ?= $(TOP_DIR)/build/bin
-LIB_DIR ?= $(TOP_DIR)/build/${TARGET}/lib
+LIB_DIR ?= $(TOP_DIR)/build/$(TARGET)/lib
 BUILD_DIR ?= $(TOP_DIR)/build/$(TARGET)/obj
 
 # Non-Moveable Directories
 FRAMEWORK_DIR = $(UT_CORE_DIR)/framework
 UT_CONTROL = $(FRAMEWORK_DIR)/ut-control
 
+# Flags
 XCFLAGS := $(KCFLAGS)
-
-# Enable CUnit Requirements
-XCFLAGS += -DUT_CUNIT
-CUNIT_DIR +=  $(FRAMEWORK_DIR)/CUnit-2.1-3/CUnit
-CUNIT_SRC_DIRS += $(CUNIT_DIR)/Sources
-INC_DIRS += $(CUNIT_DIR)/Headers
-SRC_DIRS += $(CUNIT_SRC_DIRS)/Framework
-
-INC_DIRS += $(UT_CORE_DIR)/include
-INC_DIRS += $(UT_CONTROL)/include
-INC_DIRS += $(UT_CORE_DIR)/src
-
-SRC_DIRS += $(UT_CORE_DIR)/src
-
-XLDFLAGS += -L $(UT_CONTROL)/build/$(TARGET)/lib -lut_control
-
-MKDIR_P ?= @mkdir -p
-
-# Defaults for target linux
-ifeq ($(TARGET),linux)
-CUNIT_VARIANT=i686-pc-linux-gnu
-CC := gcc -ggdb -o0 -Wall
+INC_DIRS += $(UT_CORE_DIR)/include $(UT_CONTROL)/include $(UT_CORE_DIR)/src
+
+# Set up compiler and source directories based on DGTEST
+ifneq ($(DGTEST),1) # CUNIT case
+  TARGET_EXEC ?= hal_test
+  COMPILER := $(if $(filter arm,$(TARGET)),$(CC),gcc)
+  DGTEST = 0
+
+  # Enable CUnit Requirements
+  XCFLAGS += -DUT_CUNIT
+  CUNIT_DIR = $(FRAMEWORK_DIR)/CUnit-2.1-3/CUnit
+  INC_DIRS += $(CUNIT_DIR)/Headers $(UT_CORE_DIR)/src/cunit
+  SRC_DIRS += $(CUNIT_DIR)/Sources/Framework $(UT_CORE_DIR)/src
+  XLDFLAGS += $(YLDFLAGS) $(LDFLAGS) -L$(UT_CONTROL)/build/$(TARGET)/lib -lut_control -Wl,-rpath, -pthread -lpthread
+
+  # Source files
+  SRCS := $(shell find $(SRC_DIRS) -name *.c -or -name *.s)
+else # GTEST case
+  TARGET_EXEC ?= hal_gtest
+  COMPILER := $(if $(filter arm,$(TARGET)),$(CXX),g++)
+  DGTEST = 1
+
+  SRC_DIRS += $(UT_CORE_DIR)/src
+  EXCLUDE_DIRS = $(SRCDIR)/cunit
```

Comment:
u shouldn't need to exclude anything you should have not included it

Reply:
so, our src dir structure is like this, as of today with this change:
```
src/
├── cpp_source
├── c_source
│   ├── cunit_lgpl
│   │   ├── COPYING
│   │   ├── COPYING.LGPL
│   │   ├── patches
│   │   │   └── CorrectBuildWarningsInCunit.patch
│   │   ├── ut_automated.c
│   │   ├── ut_basic.c
│   │   └── ut_console.c
│   ├── ut_cunit.c
│   └── ut_cunit_internal.h
├── ut_internal.h
├── ut_kvp_profile.c
└── ut_main.c
``` 

For VARIANT=CPP, I don't want to have CUNIT sources which resides in c_source dir. However, there are few .c files that I need which are outside c_source dir. Hence I have put this exclude logic. 
Alternate, approach can be to create a common dir and give those specific dir names to the variable. In that scenario I can remove exclude logic. 
Let me know, which is better logic and I shall do changes accordingly.